In [1]:
import os
import pandas as pd
import numpy as np
import yfinance as yf
import yahoo_fin.stock_info as sf
import matplotlib.pyplot as plt
import datetime as dt
import pandas_datareader as pdr
from functools import reduce

In [31]:
pd.options.display.float_format = '{:,.2f}'.format

In [5]:
pd.set_option=('display.max_columns',None)
pd.set_option=('display.rows',None)

### Stocks from the same industry: Technology
* Apple = AAPL
* Samsung = 005930.KS
* Hon Hai Precision Industry Co = HNHPF
* Microsoft Corp = MSFT
* Dell Technologies = DELL
* Sony Corp = SNEJF
* IBM = IBM
* Intel Corp = INTC
* Pansonic Corp = PCRFY
* HP = HPQ

In [6]:
start = dt.datetime(2012,3,6)
end = dt.datetime(2022,3,6)

In [7]:
def get_stock(ticker):
    data=pdr.DataReader(f"{ticker}","yahoo",start,end)
    data[f"{ticker}"]=data["Adj Close"]
    data=data[[f'{ticker}']]
    print(data.head())
    return data

In [9]:
apple = get_stock("AAPL")
samsung = get_stock("005930.KS")
hon = get_stock("HNHPF")
microsoft = get_stock("MSFT")
dell = get_stock("DELL")
sony = get_stock("SNEJF")
ibm = get_stock("IBM")
intel = get_stock("INTC")
pansonic = get_stock("PCRFY")
hp = get_stock("HPQ")

            AAPL
Date            
2012-03-06 16.22
2012-03-07 16.23
2012-03-08 16.57
2012-03-09 16.67
2012-03-12 16.88
            005930.KS
Date                 
2012-03-06  19,219.55
2012-03-07  19,089.25
2012-03-08  19,219.55
2012-03-09  20,033.94
2012-03-12  19,708.18
            HNHPF
Date             
2012-03-06   3.59
2012-03-07   3.48
2012-03-08   3.62
2012-03-09   3.61
2012-03-12   3.77
            MSFT
Date            
2012-03-06 25.66
2012-03-07 25.89
2012-03-08 26.03
2012-03-09 26.01
2012-03-12 26.05
            DELL
Date            
2016-08-17 12.07
2016-08-18 12.00
2016-08-19 12.21
2016-08-22 12.25
2016-08-23 12.63
            SNEJF
Date             
2012-03-06  19.71
2012-03-07  19.75
2012-03-08  19.75
2012-03-09  20.45
2012-03-12  20.22
              IBM
Date             
2012-03-06 130.67
2012-03-07 131.01
2012-03-08 132.36
2012-03-09 132.90
2012-03-12 133.15
            INTC
Date            
2012-03-06 19.84
2012-03-07 20.06
2012-03-08 20.01
2012-03-09 20.18
2012-03-1

In [14]:
def merge_stocks(tickers):
    data_frame = []
    for i in tickers:
            data_frame.append(get_stock(i))
    df_merged = reduce(lambda left, right:pd.merge(left,right,on=["Date"],how="outer"),data_frame)
    print(df_merged.head())
    return df_merged

In [15]:
techstocks = ['AAPL','005930.KS','MSFT','DELL','SNEJF','IBM','INTC','PCRFY','HPQ']

In [16]:
merge_stocks(techstocks)

            AAPL
Date            
2012-03-06 16.22
2012-03-07 16.23
2012-03-08 16.57
2012-03-09 16.67
2012-03-12 16.88
            005930.KS
Date                 
2012-03-06  19,219.55
2012-03-07  19,089.24
2012-03-08  19,219.55
2012-03-09  20,033.94
2012-03-12  19,708.18
            MSFT
Date            
2012-03-06 25.66
2012-03-07 25.89
2012-03-08 26.03
2012-03-09 26.01
2012-03-12 26.05
            DELL
Date            
2016-08-17 12.07
2016-08-18 12.00
2016-08-19 12.21
2016-08-22 12.25
2016-08-23 12.63
            SNEJF
Date             
2012-03-06  19.71
2012-03-07  19.75
2012-03-08  19.75
2012-03-09  20.45
2012-03-12  20.22
              IBM
Date             
2012-03-06 130.67
2012-03-07 131.01
2012-03-08 132.36
2012-03-09 132.90
2012-03-12 133.15
            INTC
Date            
2012-03-06 19.84
2012-03-07 20.06
2012-03-08 20.01
2012-03-09 20.18
2012-03-12 20.12
            PCRFY
Date             
2012-03-06   8.58
2012-03-07   8.69
2012-03-08   8.78
2012-03-09   8.97
2012-03-12

AAPL  005930.KS  MSFT  DELL  SNEJF    IBM  INTC  PCRFY  HPQ
Date                                                                   
2012-03-06 16.22  19,219.55 25.66   NaN  19.71 130.67 19.84   8.58 8.27
2012-03-07 16.23  19,089.24 25.89   NaN  19.75 131.01 20.06   8.69 8.33
2012-03-08 16.57  19,219.55 26.03   NaN  19.75 132.36 20.01   8.78 8.43
2012-03-09 16.67  20,033.94 26.01   NaN  20.45 132.90 20.18   8.97 8.27
2012-03-12 16.88  19,708.18 26.05   NaN  20.22 133.15 20.12   8.93 8.26
...          ...        ...   ...   ...    ...    ...   ...    ...  ...
2021-09-06   NaN  76,588.41   NaN   NaN    NaN    NaN   NaN    NaN  NaN
2021-11-25   NaN  73,368.67   NaN   NaN    NaN    NaN   NaN    NaN  NaN
2021-12-24   NaN  80,138.10   NaN   NaN    NaN    NaN   NaN    NaN  NaN
2022-01-17   NaN  77,500.00   NaN   NaN    NaN    NaN   NaN    NaN  NaN
2022-02-21   NaN  74,300.00   NaN   NaN    NaN    NaN   NaN    NaN  NaN

[2591 rows x 9 columns]

In [17]:
techport = merge_stocks(techstocks)

            AAPL
Date            
2012-03-06 16.22
2012-03-07 16.23
2012-03-08 16.57
2012-03-09 16.67
2012-03-12 16.88
            005930.KS
Date                 
2012-03-06  19,219.55
2012-03-07  19,089.24
2012-03-08  19,219.55
2012-03-09  20,033.93
2012-03-12  19,708.19
            MSFT
Date            
2012-03-06 25.66
2012-03-07 25.89
2012-03-08 26.03
2012-03-09 26.01
2012-03-12 26.05
            DELL
Date            
2016-08-17 12.07
2016-08-18 12.00
2016-08-19 12.21
2016-08-22 12.25
2016-08-23 12.63
            SNEJF
Date             
2012-03-06  19.71
2012-03-07  19.75
2012-03-08  19.75
2012-03-09  20.45
2012-03-12  20.22
              IBM
Date             
2012-03-06 130.67
2012-03-07 131.01
2012-03-08 132.36
2012-03-09 132.90
2012-03-12 133.15
            INTC
Date            
2012-03-06 19.84
2012-03-07 20.06
2012-03-08 20.01
2012-03-09 20.18
2012-03-12 20.12
            PCRFY
Date             
2012-03-06   8.58
2012-03-07   8.69
2012-03-08   8.78
2012-03-09   8.97
2012-03-12

In [18]:
techport.shape

(2591, 9)

In [20]:
from pypfopt.expected_returns import mean_historical_return
from pypfopt.risk_models import CovarianceShrinkage

In [21]:
mean = mean_historical_return(techport)
s = CovarianceShrinkage(techport).ledoit_wolf()

In [22]:
from pypfopt.efficient_frontier import EfficientFrontier

In [23]:
EF = EfficientFrontier(mean,s)
weights = EF.max_sharpe()

In [24]:
clean_weights = EF.clean_weights()
print(dict(clean_weights))

{'AAPL': 0.14202, '005930.KS': 0.07881, 'MSFT': 0.29101, 'DELL': 0.40521, 'SNEJF': 0.08296, 'IBM': 0.0, 'INTC': 0.0, 'PCRFY': 0.0, 'HPQ': 0.0}


In [27]:
EF.portfolio_performance(verbose=True)

Expected annual return: 25.4%
Annual volatility: 18.3%
Sharpe Ratio: 1.27


(0.2538005321176163, 0.1834472000839973, 1.2744840586858945)

In [28]:
from pypfopt.discrete_allocation import DiscreteAllocation,get_latest_prices

In [33]:
new_prices = get_latest_prices(techport)
DA = DiscreteAllocation(weights,new_prices,total_portfolio_value=10000)

allocation,leftover=DA.greedy_portfolio()

print("Discrete Allocation:",allocation)
print("Funds Remaining: ${:.2f}".format(leftover))

Discrete Allocation: {'DELL': 77, 'MSFT': 10, 'AAPL': 8, 'SNEJF': 8}
Funds Remaining: $987.12


### Stocks from different industries:
* Dell = DELL (Technology)
* Lamar Advertising Company = LAMR (Advertising)
* iShares US Aerospace and Defense ETF = ITA (Aerospace)
* Invesco DB Agricultural Fund = DBA (Agriculture)
* Construction Partners = ROAD (Construction)
* TAL Education Group = TAL (Education)
* Houston America Energy Corp = HUSA (Energy)
* AMC Entertainment Holdings = AMC (Entertainment)
* Macy's = M (Fashion)
* Tyson Foods = TSN (Food)


In [37]:
macys = get_stock("M")
amc = get_stock("AMC")
husa = get_stock("HUSA")
tal = get_stock("TAL")
road = get_stock("ROAD")
dba = get_stock("DBA")
ita = get_stock("ITA")
lamar = get_stock("LAMR")
dell = get_stock("DELL")
tyson = get_stock("TSN")

               M
Date            
2012-03-06 25.87
2012-03-07 26.70
2012-03-08 27.27
2012-03-09 27.53
2012-03-12 27.47
             AMC
Date            
2013-12-18 13.59
2013-12-19 14.02
2013-12-20 14.15
2013-12-23 14.20
2013-12-24 14.05
            HUSA
Date            
2012-03-06 81.50
2012-03-07 83.12
2012-03-08 77.62
2012-03-09 80.25
2012-03-12 77.38
            TAL
Date           
2012-03-06 1.79
2012-03-07 1.76
2012-03-08 1.80
2012-03-09 1.79
2012-03-12 1.77
            ROAD
Date            
2018-05-04 12.10
2018-05-07 12.05
2018-05-08 12.05
2018-05-09 12.02
2018-05-10 12.00
             DBA
Date            
2012-03-06 27.90
2012-03-07 27.66
2012-03-08 27.74
2012-03-09 27.67
2012-03-12 27.78
             ITA
Date            
2012-03-06 28.73
2012-03-07 28.92
2012-03-08 29.34
2012-03-09 29.51
2012-03-12 29.50
            LAMR
Date            
2012-03-06 21.48
2012-03-07 21.52
2012-03-08 21.79
2012-03-09 21.70
2012-03-12 21.69
            DELL
Date            
2016-08-17 12.07
2016

In [38]:
divstocks = ['M','AMC','HUSA','TAL','ROAD','DBA','ITA','LAMR','DELL','TSN']

In [39]:
merge_stocks(divstocks)

               M
Date            
2012-03-06 25.87
2012-03-07 26.70
2012-03-08 27.27
2012-03-09 27.53
2012-03-12 27.47
             AMC
Date            
2013-12-18 13.59
2013-12-19 14.02
2013-12-20 14.15
2013-12-23 14.20
2013-12-24 14.05
            HUSA
Date            
2012-03-06 81.50
2012-03-07 83.12
2012-03-08 77.62
2012-03-09 80.25
2012-03-12 77.38
            TAL
Date           
2012-03-06 1.79
2012-03-07 1.76
2012-03-08 1.80
2012-03-09 1.79
2012-03-12 1.77
            ROAD
Date            
2018-05-04 12.10
2018-05-07 12.05
2018-05-08 12.05
2018-05-09 12.02
2018-05-10 12.00
             DBA
Date            
2012-03-06 27.90
2012-03-07 27.66
2012-03-08 27.74
2012-03-09 27.67
2012-03-12 27.78
             ITA
Date            
2012-03-06 28.73
2012-03-07 28.92
2012-03-08 29.34
2012-03-09 29.51
2012-03-12 29.50
            LAMR
Date            
2012-03-06 21.48
2012-03-07 21.52
2012-03-08 21.79
2012-03-09 21.70
2012-03-12 21.69
            DELL
Date            
2016-08-17 12.07
2016

M   AMC  HUSA  TAL  ROAD   DBA    ITA   LAMR  DELL   TSN
Date                                                                   
2012-03-06 25.87   NaN 81.50 1.79   NaN 27.90  28.73  21.48   NaN 16.37
2012-03-07 26.70   NaN 83.12 1.76   NaN 27.66  28.92  21.52   NaN 16.69
2012-03-08 27.27   NaN 77.62 1.80   NaN 27.74  29.34  21.79   NaN 16.77
2012-03-09 27.53   NaN 80.25 1.79   NaN 27.67  29.51  21.70   NaN 16.95
2012-03-12 27.47   NaN 77.38 1.77   NaN 27.78  29.50  21.69   NaN 16.79
...          ...   ...   ...  ...   ...   ...    ...    ...   ...   ...
2022-02-28 25.92 18.86  1.41 2.84 26.76 20.93 111.90 109.06 50.96 92.66
2022-03-01 24.53 18.32  1.47 2.89 25.25 21.44 110.71 107.17 49.88 93.83
2022-03-02 26.19 18.53  1.47 2.70 25.63 21.43 111.54 109.49 53.60 95.04
2022-03-03 26.03 18.06  1.43 2.61 26.66 21.91 109.85 110.41 52.59 95.34
2022-03-04 24.80 16.57  1.90 2.37 25.90 21.84 109.33 107.62 51.96 93.51

[2517 rows x 10 columns]

In [40]:
divport = merge_stocks(divstocks)

               M
Date            
2012-03-06 25.87
2012-03-07 26.70
2012-03-08 27.27
2012-03-09 27.53
2012-03-12 27.47
             AMC
Date            
2013-12-18 13.59
2013-12-19 14.02
2013-12-20 14.15
2013-12-23 14.20
2013-12-24 14.05
            HUSA
Date            
2012-03-06 81.50
2012-03-07 83.12
2012-03-08 77.62
2012-03-09 80.25
2012-03-12 77.38
            TAL
Date           
2012-03-06 1.79
2012-03-07 1.76
2012-03-08 1.80
2012-03-09 1.79
2012-03-12 1.77
            ROAD
Date            
2018-05-04 12.10
2018-05-07 12.05
2018-05-08 12.05
2018-05-09 12.02
2018-05-10 12.00
             DBA
Date            
2012-03-06 27.90
2012-03-07 27.66
2012-03-08 27.74
2012-03-09 27.67
2012-03-12 27.78
             ITA
Date            
2012-03-06 28.73
2012-03-07 28.92
2012-03-08 29.34
2012-03-09 29.51
2012-03-12 29.50
            LAMR
Date            
2012-03-06 21.48
2012-03-07 21.52
2012-03-08 21.79
2012-03-09 21.70
2012-03-12 21.69
            DELL
Date            
2016-08-17 12.07
2016

In [41]:
divport.shape

(2517, 10)

In [42]:
mean2 = mean_historical_return(divport)
s2 = CovarianceShrinkage(divport).ledoit_wolf()

In [43]:
EF2 = EfficientFrontier(mean2,s2)
weights2 = EF2.max_sharpe()

In [45]:
clean_weights2 = EF.clean_weights()
print(dict(clean_weights2))

{'AAPL': 0.14202, '005930.KS': 0.07881, 'MSFT': 0.29101, 'DELL': 0.40521, 'SNEJF': 0.08296, 'IBM': 0.0, 'INTC': 0.0, 'PCRFY': 0.0, 'HPQ': 0.0}


In [47]:
EF2.portfolio_performance(verbose=True)

Expected annual return: 23.4%
Annual volatility: 23.7%
Sharpe Ratio: 0.90


(0.23393269184569873, 0.23655701511564856, 0.9043599562714757)

In [49]:
latest_price = get_latest_prices(divport)
DA2 = DiscreteAllocation(weights2,latest_price,total_portfolio_value=10000)

allocation,leftover=DA2.greedy_portfolio()

print("Discrete Allocation:",allocation)
print("Funds Remaining: ${:.2f}".format(leftover))

Discrete Allocation: {'DELL': 75, 'ROAD': 82, 'TSN': 21, 'ITA': 9, 'LAMR': 9}
Funds Remaining: $62.94


#### Our portfolio with 10 stocks from the technology industry, we got these results with the Discrete Allocation: 
Discrete Allocation: {'DELL': 77, 'MSFT': 10, 'AAPL': 8, 'SNEJF': 8}
Funds Remaining: $987.12

For the undiversified portfolio, our expected annual return was 25.4% with an annual volatility of 18.3% and a Sharpe Ratio of 1.27 which is okay.

#### Our portfolio with diversified stock, we got these results with the Discrete Allocation:
Discrete Allocation: {'DELL': 75, 'ROAD': 82, 'TSN': 21, 'ITA': 9, 'LAMR': 9}
Funds Remaining: $62.94

For our diversified portfolio, our expected annual return rate was 23.4% (which is less than the technology only portfolio), the annual volatility was 23.7% (which is significantly higher than our technology portfolio), and a Sharpe Ratio of 0.90 (which is worse than our technology portfolio). 
I think the reason for these numbers has to do with the certain stocks I chose. I might have chosen some stocks that were not good enough to compete against some of the best technology stocks there are. 